give.php

```diff
@@ -133,6 +134,7 @@ final class Give {
 		LegacyServiceProvider::class,
 		RestAPI::class,
 		Onboarding::class,
+		Milestones::class,
```

Comment:
@JasonTheAdams Nitpick regarding naming... Since WP core has dropped references to _Gutenberg_ in favor of _Block Editor_, I think we should follow suit and just call that namespace `Blocks` rather than `GutenbergBlocks`. Does that work?

Reply:
I was actually going back and forth on that. I was debating between a generic `Components` root directory for combining all front-end components, versus discreet block and shortcode directories:
```
.
└── src
    └── Components
        └── Milestones
            └── Block
            └── Shortcode

.
└── src
    ├── Blocks
    │   └── Milestones
    └── Shortcodes
        └── Milestones
```

Reply:
I like Option #1 because a Block and a Shortcode should ideally be two different means of rendering the same component. I'd think there would be other file(s) for the shared functionality of the component itself.

With Option #2, it's not clear to me where that shared functionality would live.

Reply:
Hmm...was mainly thinking in terms of the "feature first/domain based" architecture we had discussed moving toward a few weeks back. In my opinion, Milestones will ultimately grow in complexity beyond a simple block registration (especially once we need to introduce logic related to aggregating totals across forms for several different metrics). I definitely think your recommendation makes a lot of sense, just that we should pick an approach and stick with it, since we currently kinda walk a funky line where, for example, REST API routes exist both in the REST API ServiceProvider _and_ in the Onboarding service provider. Based on our conversation with @kjohnson, I had been given the impression that the approach we took with Onboarding was correct, and that Reports API routes would ultimately be refactored into a Reports ServiceProvider. 

Let me know if I'm completely off here, very open to your approach cause it makes sense to me - but wanna make sure I understand what our approach with ServiceProviders is exactly.

Reply:
The contrast between the REST API controllers in Onboarding vs Reports is a good example here - MVC architecture doesn't have to only live at the top level of the project. Each domain can organize its own models, views, and controllers.

It might be easy to find a block if all the blocks are in the same place, but we don't work this way. This is why I prefer for these decisions to happen organically, as opposed to strictly-theoretical decisions. At first glance it is easy to say "put the blocks here", "put the REST controllers there", but again, in practice this increases complexity of the work at hand. 

I would prefer to self-contain where possible and then possibly extract or "promote" where necessary. In this case, the block belongs to the domain - as opposed to the framework. We are layering in new functionality.

That said, I think it does make sense to compile, or "publish" all built assets to an organized location - but this can be automated. We can work in packages and publish to assets.

I'm fine with splitting the "assets" and the "src" if that is the decision, so

```
assets
  |--> Milestones
src
  |--> Milestones
```

... but am not a huge fan of working out of split directories, as opposed to

```
assets
  |--> dist
      |--> milestones.css
src
  |--> Milestones
      | --> assets
      |--> index.php
```

.. which is more of a package approach.

Reply:
@kjohnson Next time you're around please let me know. I'd like to hop on a call to talk over this.

Reply:
Following up on this conversation, here's a link to a draft PR that introduces logic for displaying aggregated revenue in a Milestone block: https://github.com/impress-org/givewp/pull/5236

In that PR, because the complexity of Milestones grew, you can see that I introduced a Milestone model which handles all logic essential to calculating Milestone progress and creating Milestone output. This allows the Block class in the milestones directory to be solely concerned with registering the Milestone block, and opens the door for creating a Shortcode class to handle registering the Milestone shortcode.